Review PR: Proposed text for subsampled coordinates (#1)
erget/cf-conventions: appj.adoc

[[appendix-coordinate-sampling, Appendix J, Coordinate Sampling]] 
[appendix]
== Coordinate Sampling

> The definitions and guidance given here allow an application to compress an existing data set using coordinate sampling, while letting the creator of the compressed dataset control the accuracy of the reconstituted coordinates through the degree of subsambling and the choice of interpolation method.

Reviewer: Small typo
```suggestion
The definitions and guidance given here allow an application to compress an existing data set using coordinate sampling, while letting the creator of the compressed dataset control the accuracy of the reconstituted coordinates through the degree of subsampling and the choice of interpolation method.
```

In the case of two dimensional interpolation, the two interpolation variables are equivalently computed as

`s1 = s1(i) = (i1 - ia1)/(ib1 - ia1)`  +
`s2 = s2(i) = (i2 - ia2)/(id2 - ia2)`

where `ia1` and `ib1` are the first dimension indices in the target domain of the tie points `A` and `B` respectively, `ia2` and `id2` are the second dimension indices in the target domain of the tie points `A` and `D respectively and the indices `i1` and `i2` are the first and second dimension indices respectively in the target domain of the coordinate value to be reconstituted. 

For the reconstitution of the uncompressed coordinate and auxiliary coordinate variables the interpolation method can be applied independently for each interpolation zone, making it possible to parallelize the computational process.

The following notation is used: +  
For a vector `v`,  `v.x` , `x.y` and `v.z` refer to the three coordinates of that vector. +
For a coordinate pair `ll`, `ll.lat` and `ll.lon` refer to the latitude and longitude coordinates. +

[[coordinate_conversion]]
==== Common conversions and formulas

[cols="8,12"] 
|=============== 

> | Description | Formula

Reviewer: I think adding a column for function names (fll2xyz, fcross, norm, etc...) would make it easier for users to find function descriptions when they encounter the function names elsewhere in the document.

Reviewer 2: Could we achieve that by adding the function name in bold as the first line in existing Description column?
Like in 

**fll2xyz()**
Conversion from geocentric (latitude, longitude) to cartesian vector (x, y, z)

| Calculate the linear interpolation value of one coordinate `u` in one dimension between tie points `a` and `b`   | `u = fl1(ua, ub, s) = (1 - s)*ua + s*u2`; + 

| Calculate the quadratic interpolation value of one coordinate `u` in one dimension between tie points `a` and `b` with interpolation coefficient `c`   | `u = fq1(ua, ub, c, s) = (1 - s)*ua + (1 - s)*s1*c + s*u2`; + 

| Calculate the quadratic interpolation values of two coordinates `(u.1, u.2)` in one dimension between tie points `a` and `b` with interpolation coefficients `(c.1, c.2)`  | `(u.1, u.2) = fq2(ua, ub, c, s) = (fq1(ua.1, ua.2, c.1, s), fq1(ua.1, ua.2, c.1, s))``; +

| Calculate the quadratic interpolation values of three coordinates `(u.1, u.2, u.3)` in one dimension between tie points `a` and `b` with interpolation coefficients `(c.1, c.2, c.3)`  | `(u.1, u.2, u.3) = fq3(ua, ub, c, s) = (fq1(ua.1, ub.1, c.1, s), fq1(ua.2, ub.2, c.2, s), fq1(ua.3 ub.3, c.3, s))`; +

> | Vector Sum | `(x, y, z) = va + vb = (va.x + vb.x, va.y + vb.y, va.z + vb.z)` + 

Reviewer: Function name should be defined for consistency (and it is referred to in other places)
```suggestion
| Vector Sum | `(x, y, z) = fplus(va, vb) = va + vb = (va.x + vb.x, va.y + vb.y, va.z + vb.z)` + 
```

Reviewer 2: OK, agreed.

> | Vector Difference | `(x, y, z) = va - vb = (va.x - vb.x, va.y - vb.y, va.z - vb.z)` + 

Reviewer: Function name should be defined for consistency (and it is referred to in other places)
```suggestion
| Vector Difference | `(x, y, z) = fminus(va, vb) = va - vb = (va.x - vb.x, va.y - vb.y, va.z - vb.z)` + 
```

Reviewer 2: OK, agreed

 
| Vector Cross Product | `(x, y, z) = fcross(va, vb) = (va.y*vb.z - va.z*vb.y, va.z*vb.x - vva.zx*b.z, va.x*vb.y - va.y*vb.x)` + 

| Normalised Vector | `(x, y, z) = norm(v) = (v.x, v.y, v.z) / sqrt (v.x*v.x + v.y*v.y + v.z*v.z)` + 

| Vector Dot Product | `d = fdot(va, vb) = va.x*vb.x + va.y*vb.y, va.z*vb.z`

> | Vector from ce, ca  | `(x, y, z) = fp(va, vb, ce, ca) = norm(0.5*fplus(va, vb) + ce*fminus(vb, va) + ca*fcross(vb, va))`

Reviewer: Adding a defintion/description for ce, ca might be necessary

Reviewer 2: Yes, that is needed. Let's discuss where to put it.

[[interpolation_methods]]
=== Interpolation Methods

==== Linear Interpolation

[cols="6,15"]
|===============
| Name | **`interpolation_name = "linear"`** 
| Description | General purpose one dimensional linear interpolation method for one or more coordinates
| Interpolation Parameter terms  | None
| Coordinate Compression Calculations | None
| Coordinate Uncompression Calculations | 
 The coordinate value `u(i)` at index `i` between tie points `a` and `b` is calculated from:   +
 `u(i) = fl(ua, ub, s(i)) = (1 - s)*ua + s*ub`; + 
where `ua` and `ub` are the coordinate values at tie points `a` and `b` respectively. +

|===============

==== Bilinear Interpolation 

[cols="6,15"]
|===============
| Name | **`interpolation_name = "bi_linear"`** 
| Description | General purpose two dimensional linear interpolation method for one or more coordinates
| Interpolation Parameter terms | None
| Coordinate Compression Calculations | None
| Coordinate Uncompression Calculations | 

> The interpolation function fl() defined for linear interpolation above is first applied twice in the interpolation dimension 1, once between between tie points `a` and `b` and once between tie point `c` and `d`. It is then applied once in the interpolation dimension 2, between the two resulting coordinate points, yielding the interpolated coordinate value `u(i1, i2)`:  +  

Reviewer: typo
```suggestion
The interpolation function fl() defined for linear interpolation above is first applied twice in the interpolation dimension 1, once between tie points `a` and `b` and once between tie points `c` and `d`. It is then applied once in the interpolation dimension 2, between the two resulting coordinate points, yielding the interpolated coordinate value `u(i1, i2)`:  +  
```

[[quadratic2, figure 3]]
[.text-center]
.With the expansion coefficient ce > 0 and the alignment coefficient ca > 0, the method reconstitutes the points at intervals of expanding size (ce) along an arc with an alignment offset (ca) from the great circle between tie points A and B.
image::images/ci_quadratic2.svg[,100%,pdfwidth=50vw,align="center"]  

[cols="6,15"]
|===============
| Name | **`interpolation_name = "quadratic_geo"`** 
| Description | A one dimensional quadratic method for interpolation of the geographic coordinates latitude and longitude, typically used for remote sensing products with geographic coordinates on the reference ellipsoid. Requires a coordinate pair with `standard_name` `latitude` and `longitude`. +
  + 
By default, interpolation is performed directly in the latitude and longitude coordinates, but may be performed in cartesian coordinates where required for achieving the desired accuracy. This must be indicated by setting the `location_use_cartesian` flag within the interpolation parameter `interpolation_zone_flags` for each interpolation zone where interpolation in cartesian coordinates is required.  

> The quadratic interpolation coefficients `ce` and `ca`, stored as interpolation parameters in the product, describes a point `P` between the tie points `A` and `B`, which is equivalent of an additional tie point in the sense that the method will accurately reconstitute the point `P` in the same way as it accurately reconstitutes the tie points `A` and `B`. See <<quadratic1>> and <<quadratic2>>.

Reviewer: typo
```suggestion
The quadratic interpolation coefficients `ce` and `ca`, stored as interpolation parameters in the product, describe a point `P` between the tie points `A` and `B`, which is equivalent of an additional tie point in the sense that the method will accurately reconstitute the point `P` in the same way as it accurately reconstitutes the tie points `A` and `B`. See <<quadratic1>> and <<quadratic2>>.
```

Although equivalent to a tie point, the coefficients `ce` and `ca` have two advantages over tie points. Firstly, they can often be stored as a lower precision floating point number compared to the tie points, as `ce` and `ca` only describes the position of P relative to the midpoint `M` between the tie points `A` and `B`. Secondly, if any of `ce` and `ca` do not contribute significantly to the accuracy of the reconstituted points, it can be left out of the data set and its value will default to zero during uncompression. 

The coefficients may be represented in three different ways:

> For storage in the dataset as the non-dimensional coefficients `(ce, ca)`, referred to as the parametric representation. The component `ce` is the offset projected on the line from tie point `b` to tie point `a` and expressed as a fraction of the distance between `A` and `B`. The component `ca` is the offset projected on the line perpendicular to the line from tie point `b` to tie point `a` and  perpendicular to the plane spanned `va` and `vb`, the vector representations of the two tie points, and expressed as a fraction of the distance between `A` an `B`. +

Reviewer: typo
```suggestion
For storage in the dataset as the non-dimensional coefficients `(ce, ca)`, referred to as the parametric representation. The component `ce` is the offset projected on the line from tie point `b` to tie point `a` and expressed as a fraction of the distance between `A` and `B`. The component `ca` is the offset projected on the line perpendicular to the line from tie point `b` to tie point `a` and  perpendicular to the plane spanned by `va` and `vb`, the vector representations of the two tie points, and expressed as a fraction of the distance between `A` and `B`. +
```

`rsqr = fdot(vr, vr)`. +
`cr = fsqrt(1 - ce(iz)*ce(iz) - ca(iz)*ca(iz)) - fsqrt(rsqr)`; +
If the flag `location_use_cartesian` of the interpolation parameter term `interpolation_zone_flags(iz2, iz1)` is set, use the following expression to reconstitute any point `p(i)` between the tie points `a` and `b` using interpolation in cartesian coordinates + 
`vp(i) = fqxyz(va, vb, vcxyz, s(i)) = (fq(va.x, vb.x, vcxyz.x, s(i)), fq(va.y, vb.y, vcxyz.y, s(i)), fq(va.z, vb.z, vcxyz.z, s(i)))` +
Otherwise, first calculate latitude-longitude representation of the interpolation coefficients + 
`vcll = fcxyz2cll(va, vb, vcea) = (fc(a.lat, b.lat, pll.lat, 0.5), fc(a.lon, b.lon, pll.lon, 0.5))` + 
where `pll = fxyz2ll(fqxyz(va, vb, vcxyz, 0.5))`. +
Then use the following expression to reconstitute any point `p(i)` between the tie points `a` and `b` using interpolation in latitude-longitude coordinates + 
`p(i) = (p(i).lat, p(i).lon) = fqll(va, vb, vcll, s(i)) = ((fq(a.lat, b.lat, vcll.lat, s(i)), fq(a.lon, b.lon, .vcll.lon, s(i))`. + 
|===============   
  

[[bi_quadratic_geo]]
==== Biquadratic Interpolation of geographic coordinates

[[quadratic3, figure 4]]
[.text-center]

> .The the parametric representation of the interpolation coefficients `(ce, ca)`, stored in the interpolation parameters `ce1, ca1, ce2, ca2, ce3` and `ca3`, are equivalent to five additional tie points for the interpolation zone. Shown with parameter orientation and indices.  

Reviewer: typo
```suggestion
.The parametric representation of the interpolation coefficients `(ce, ca)`, stored in the interpolation parameters `ce1, ca1, ce2, ca2, ce3` and `ca3`, is equivalent to five additional tie points for the interpolation zone. Shown with parameter orientation and indices.  
```

image::images/ci_quadratic3.svg[,50%,pdfwidth=50vw,align="center"] 

[cols="6,15"]
|===============
| Name | **`interpolation_name = "bi_quadratic_geo"`** 

> | Description | A two dimensional quadratic method for interpolation of the geographic coordinates latitude and longitude, typically used fro remote sensing products with geographic coordinates on the reference ellipsoid. Requires a coordinate pair with `standard_name` `latitude` and `longitude`. +

Reviewer: typo
```suggestion
| Description | A two dimensional quadratic method for interpolation of the geographic coordinates latitude and longitude, typically used for remote sensing products with geographic coordinates on the reference ellipsoid. Requires a coordinate pair with `standard_name` `latitude` and `longitude`. +
```

The functions `fcxyz()`, `fcxyz2cea()`, `fcea2cxyz()`, `fcxyz2cll()`, `fqxyz()` and `fqll()` referenced in the following are defined in <<quadratic_geo>>. As for that method, interpolation is performed directly in the latitude and longitude coordinates or in cartesian coordinates, where required for achieving the desired accuracy. Similarly, it shares the three different representations of the quadratic interpolation coefficients, the parametric representation `(ce, ca)` for storage in the dataset, `vcll = (clat, clon)` for interpolation in geographic coordinates latitude and longitude and `vcxyz = (cx, cy, cz)` for cartesian interpolation. 

> The the parametric representation of the interpolation coefficients are store in the interpolation parameters `ce1, ca1, ce2, ca2, ce3` and `ca3` equivalent to five additional tie points for the interpolation zone as shown in <<quadratic3>>, which also shows the orientation and indices of the parameters.  

Reviewer: typo
```suggestion
The parametric representation of the interpolation coefficients, stored in the interpolation parameters `ce1, ca1, ce2, ca2, ce3` and `ca3`, is equivalent to five additional tie points for the interpolation zone as shown in <<quadratic3>>, which also shows the orientation and indices of the parameters.  
```

| Coordinate Compression Calculations | 

First calculate the cartesian representation of the interpolation coefficients sets from the tie points as well as a point `p(i2, i1)` between the tie points. If the size of the interpolation zone in the first dimension `(ib1 - ia1)` is an even number, then the index `i1 = (ib1 - ia1)/2` shall be selected for this calculation, otherwise the index `i1 = (ib1 - ia1 - 1)/2` shall be selected. If the size of the interpolation zone in the second dimension `(ib2 - ic2)` is an even number, then the index `i2 = (ib2 - ic2)/2` shall be selected for this calculation, otherwise the index `i2 = (ib2 - ic2 - 1)/2` shall be selected. +
Using the selected `(i2, i1)`, the cartesian interpolation coefficients are found from + 
`vcxyz_ab  = fcxyz( va, vb, vp(ia2, i1), s(i1));`   `vcxyz_cd  = fcxyz( vc, vd, vp(ic2, i1), s(i1));` + 
`vcxyz_ac  = fcxyz( va, vc, vp(i2, ia1), s(i2));`   `vcxyz_bd  = fcxyz( vb, vd, vp(i2, ib1), s(i2));` + 
`vcxyz_z  = fcxyz( fqxyz(va, vc, vcxyz_ac, 0.5), fqxyz(vb, vd, vcxyz_bd, 0.5), vp(i2, i1), s(i1));` +

> Finally, for storage in the interpolation parameter datasets, convert the coefficients to the parametric representation + 

Reviewer: ```suggestion
Finally, before storing them in the dataset's interpolation parameters, convert the coefficients to the parametric representation + 
```

For the reconstitution of the uncompressed coordinate and auxiliary coordinate variables the interpolation method can be applied independently for each interpolation zone, making it possible to parallelize the computational process.

The following notation is used: +  
A variable staring with `v` denotes a vector and `v.x` , `v.y` and `v.z` refer to the three coordinates of that vector. +
A variable staring with `ll` denotes a latitude-longitude coordinate pair and `ll.lat` and `ll.lon` refer to the latitude and longitude coordinates. +
For one dimensional interpolation, `i` is an index in the interpolation dimension, `tp` is an index in the tie point interpolation dimension and `iz` is an index in the interpolation zone dimensions. 
For two dimensional interpolation, `i1` and `i2` are indices in the interpolation dimensions, `tp1` and `tp2` are indices in the tie point interpolation dimensions and `iz1` and `iz2` are indices in the interpolation zone dimensions. +

> Note that, for simplicity of notation, the descriptions of the interpolation methods in most places leave out the indices of tie point related variables and refers to these with `a` and `b` in the one dimensional case and with `a`, `b`, `c` and `d` in the two dimensional case. In the two dimensional case, `a = tp(tp2, tp1)` then  `b = tp(tp2, tp1+1)`, `c = tp(tp2+1, tp1)` and `d = tp(tp2+1, tp1+1)` would reflect the way the tie point data would be stored in the data set, see also <<interpolation>>.

Reviewer: ```suggestion
Note that, for simplicity of notation, the descriptions of the interpolation methods in most places leave out the indices of tie point related variables and refer to these with `a` and `b` in the one dimensional case and with `a`, `b`, `c` and `d` in the two dimensional case. In the two dimensional case, `a = tp(tp2, tp1)` then  `b = tp(tp2, tp1+1)`, `c = tp(tp2+1, tp1)` and `d = tp(tp2+1, tp1+1)` would reflect the way the tie point data would be stored in the data set, see also <<interpolation>>.
```
